### PR TITLE
Issue 20155 - Allocating a struct with dtor on the GC heap can produce false pointers

### PR DESCRIPTION
**Pull request by rainers**
clear padding area not completely zeroed for struct allocations with dtor

### CODE REVIEW

**Comment by rainers** [comment]
Added fix for arrays of structs with dtor and any large arrays.

File: src/rt/lifetime.d [hunk 1]
```diff
@@ -1096,15 +1118,19 @@ extern (C) void* _d_newitemU(in TypeInfo _ti)
...
+    immutable size = itemSize + tiSize;
     if (tiSize)
         flags |= BlkAttr.STRUCTFINAL | BlkAttr.FINALIZE;
 
     auto blkInf = GC.qalloc(size, flags, ti);
     auto p = blkInf.base;
 
     if (tiSize)
+    {
+        *cast(TypeInfo*)(p + itemSize) = null; // the GC might not have cleared this area
```
Comment (kinke) [inline src/rt/lifetime.d]: Is there some guarantee that there isn't anymore padding between this and the `ti` pointer at the end of the block?

Comment (rainers) [reply src/rt/lifetime.d]: Everything else is already cleared by the GC:
https://github.com/dlang/druntime/blob/396a0ec7ef2ba38bb6314f7992d5d99be3645f02/src/gc/impl/conservative/gc.d#L417-L422

**Comment by rainers** [comment]
Added a tiny optimization that memset isn't called if the NO_SCAN attribute is set.

File: src/rt/lifetime.d [hunk 2]
```diff
@@ -436,14 +455,17 @@ BlkInfo __arrayAlloc(size_t arrsize, ref BlkInfo info, const TypeInfo ti, const
     if (!info.base)
         return __arrayAlloc(arrsize, ti, tinext);
 
+    size_t padsize = __arrayPad(arrsize, tinext);
```
Comment (Geod24) [inline src/rt/lifetime.d]: `immutable` ?

File: src/rt/lifetime.d [hunk 3]
```diff
@@ -406,6 +406,22 @@ size_t __arrayPad(size_t size, const TypeInfo tinext) nothrow pure @trusted
     return size > MAXMEDSIZE ? LARGEPAD : ((size > MAXSMALLSIZE ? MEDPAD : SMALLPAD) + structTypeInfoSize(tinext));
 }
 
+/**
+  clear padding that might not be zeroed by the GC (it assumes it is within the
+  requested size from the start, but it is actually at the end of the allocated block)
+  */
+void __arrayClearPad(ref BlkInfo info, size_t arrsize, size_t padsize) nothrow pure
```
Comment (Geod24) [inline src/rt/lifetime.d]: `private` ?

File: src/rt/lifetime.d (hunk 2 above)
Comment (rainers) [reply src/rt/lifetime.d]: done.

File: src/rt/lifetime.d (hunk 3 above)
Comment (rainers) [reply src/rt/lifetime.d]: done.

Comment (Geod24) [reply src/rt/lifetime.d]: On the wrong function?

Comment (rainers) [reply src/rt/lifetime.d]: Oops, fixed. Just shows how arbitrary it looks to mark only this function private.

**Comment by rainers** [comment]
Ping.

**Comment by ibuclaw** [comment]
This PR caused a regression https://issues.dlang.org/show_bug.cgi?id=22844

File: src/rt/lifetime.d (hunk 1 above)
Comment (ibuclaw) [reply src/rt/lifetime.d:1131]: This set is unaligned, breaking strict alignment targets.

**Pull request merged**